spec/features/admin/timeline_batch_actions_spec.rb

```diff
@@ -0,0 +1,57 @@
+require 'rails_helper'
+
+feature 'timeline batch actions', js: true do
+  let(:division) { create(:division) }
+  let(:user) { create_member(division) }
+  let!(:loan) { create(:loan, :with_timeline, division: division) }
+
+  # Create a step with a dependent step, which also has a dependent step
+  let(:step1) { create(:project_step, project: loan) }
+  let(:step2) { create(:project_step, project: loan, scheduled_start_date: nil, schedule_parent: step1) }
+  let(:step3) { create(:project_step, project: loan, scheduled_start_date: nil, schedule_parent: step2) }
+
+  # Create two independent steps, one with and the other without a scheduled start date
+  let(:step4) { create(:project_step, project: loan, scheduled_start_date: Date.today, scheduled_duration_days: nil) }
+  let(:step5) { create(:project_step, project: loan, scheduled_start_date: nil, scheduled_duration_days: nil) }
+
+  before do
+    login_as(user, scope: :user)
+  end
+
+  it 'changes dates' do
+    # Add the steps to the loan
+    loan.root_timeline_entry.children += [step1, step2, step3, step4, step5]
+
+    # Set variables
+    step1_orig_date = step1.scheduled_start_date
+    step2_orig_date = step2.scheduled_start_date
+    step3_orig_date = step3.scheduled_start_date
+    step4_orig_date = step4.scheduled_start_date
+    step5_orig_date = step5.scheduled_start_date
+    num = 3
+
+    # Select steps
+    visit(admin_loan_path(loan) + '/timeline')
+    find(:css, ".select-step[data-id='#{step1.id}']").set(true)
+    find(:css, ".select-step[data-id='#{step2.id}']").set(true)
+    find(:css, ".select-step[data-id='#{step3.id}']").set(true)
+    find(:css, ".select-step[data-id='#{step4.id}']").set(true)
+    find(:css, ".select-step[data-id='#{step5.id}']").set(true)
+
+    # Ajust dates batch action
+    click_on('Batch Actions')
+    find(:css, ".batch-actions .action.adjust-dates").click
+    fill_in('num_of_days', with: num)
+    find(:css, ".action.adjust-dates-confirm").click
+    expect(page).to have_content('successfully updated')
+
+    # Confirm dependent steps change correctly
+    expect(step1.reload.scheduled_start_date).to eq(step1_orig_date.try(:+, num))
+    expect(step2.reload.scheduled_start_date).to eq(step2_orig_date.try(:+, num))
+    expect(step3.reload.scheduled_start_date).to eq(step3_orig_date.try(:+, num))
+
+    # Confirm steps with and without start dates change correctly
+    expect(step4.reload.scheduled_start_date).to eq(step4_orig_date.try(:+, num))
+    expect(step5.reload.scheduled_start_date).to eq(step5_orig_date.try(:+, num))
+  end
+end
```

Comment:
Nice work on this feature spec!

Ideally I would like to have seen more of the logic details tested in the date_adjustment_spec, which is model spec. Usually you want to put most of the logic testing stuff in there. (Please check out the testing pyramid: https://martinfowler.com/bliki/TestPyramid.html)

Then the feature specs just need to test that the data gets into the model properly. So you wouldn't really need five steps, just a couple would do, because you would already have full faith in the date adjustment logic. You'd just be testing that the selected steps got passed into the model via the controller and the request completes successfully.

Does this make sense? It's really important to understand this distinction so please feel free to reach out if it's not clear.

In the meantime, I think this is fine for now, modulo the suggestions I made above. There is no need to redo it. We have test coverage and that's the main thing. Just flagging this for future.

Thanks for all your work!